### PR TITLE
[SPARK-31402][SQL] Fix rebasing of BCE dates/timestamps

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to fallback to rebasing via local dates/timestamps for days/micros of before common era (BCE). 

### Why are the changes needed?
It fixes the bug of rebasing dates/timestamps of BCE.

### Does this PR introduce any user-facing change?
Yes

### How was this patch tested?
- By existing tests in `RebaseDateTimeSuite` and `DateTimeUtilsSuite`
- Added tests for negative years to `RebaseDateTimeSuite`

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon Please, take a look at the fix for rebasing negative years.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon @dongjoon-hyun Please, review this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala [hunk 1]
```diff
@@ -66,15 +66,53 @@ object RebaseDateTime {
   // The starting point is the `0001-01-01` (-719164 days since the epoch in
   // Julian calendar). All dates before the staring point have the same difference
   // of 2 days in Julian and Proleptic Gregorian calendars.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala:68]: shall we remove `All dates before the staring point have the same difference ...`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala [hunk 2]
```diff
@@ -100,15 +142,46 @@ object RebaseDateTime {
   // The starting point is the `0001-01-01` (-719162 days since the epoch in
   // Proleptic Gregorian calendar). All dates before the staring point have the same
   // difference of -2 days in Proleptic Gregorian and Julian calendars.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala:144]: ditto

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala:68]: For this particular array, it is valid comment. It is not valid for `rebaseJulianToGregorianDays`, strictly speaking.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala [hunk 3]
```diff
@@ -66,15 +66,53 @@ object RebaseDateTime {
...
+
   /**
    * Converts the given number of days since the epoch day 1970-01-01 to a local date in Julian
    * calendar, interprets the result as a local date in Proleptic Gregorian calendar, and takes the
    * number of days since the epoch from the Gregorian local date.
    *
+   * @param days The number of days since the epoch in Julian calendar. It can be negative.
+   * @return The rebased number of days in Gregorian calendar.
+   */
+  private[sql] def localRebaseJulianToGregorianDays(days: Int): Int = {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/RebaseDateTime.scala:86]: nit: let's avoid to use `private[sql]`, next time.

**Pull request closed without merging**